test-suite/src/test/scala/org/scalajs/testsuite/library/UnionTypeTest.scala

```diff
@@ -0,0 +1,153 @@
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+package org.scalajs.testsuite.library
+
+import scala.language.implicitConversions
+
+import scala.scalajs.js
+import js.|
+
+import org.scalajs.jasminetest.JasmineTest
+import org.scalajs.testsuite.Typechecking._
+
+import js.JSConverters._
+
+object UnionTypeTest extends JasmineTest {
+
+  private implicit def unionAsJSAny(x: _ | _): js.Any = x.asInstanceOf[js.Any]
+
+  describe("js.| (postive)") {
+
+    it("Left and right") {
+      val x1: Int | String = 4
+      expect(x1).toBe(4)
+
+      val x2: Int | String = "hello"
+      expect(x2).toBe("hello")
+    }
+
+    it("Left and right with subtyping") {
+      val list = List(1, 2, 3)
+
+      val x1: Seq[Int] | CharSequence = list
+      expect(x1.isInstanceOf[List[_]]).toBeTruthy
+      expect(x1.asInstanceOf[List[_]].toJSArray).toEqual(js.Array(1, 2, 3))
+
+      val x2: Seq[Int] | CharSequence = "hello"
+      expect(x2).toBe("hello")
+    }
+
+    it("Three types") {
+      val x1: Int | String | Boolean = 3
+      expect(x1).toBe(3)
+
+      val x2: Int | String | Boolean = "hello"
+      expect(x2).toBe("hello")
+
+      val x3: Int | String | Boolean = false
+      expect(x3).toBe(false)
+    }
+
+    it("Upcast") {
+      val x1: List[Int] | String = "hello"
+      val x2: Seq[Int] | CharSequence = x1
+      expect(x2).toBe("hello")
+    }
+
+    it("Int as Double") {
+      val x1: Double | String = 3
+      expect(x1).toBe(3)
+    }
+
+    it("Swap base types") {
+      val x1: Int | String = 3
+      val x2: String | Int = x1
+      expect(x2).toBe(3)
+    }
+
+    it("Permutations for 3 base types") {
+      val x: Int | String | Boolean = 3
+
+      val x1: Int | Boolean | String = x
+      val x2: String | Int | Boolean = x
+      val x3: String | Boolean | Int = x
+      val x4: Boolean | Int | String = x
+      val x5: Boolean | String | Int = x
+
+      expect(x1).toBe(3)
+      expect(x2).toBe(3)
+      expect(x3).toBe(3)
+      expect(x4).toBe(3)
+      expect(x5).toBe(3)
+    }
+
+    it("Permutations of 2 base types to 3 base types") {
+      val x1: Int | String = 3
+      val x2: Int | Boolean = false
+      val x3: Boolean | String = "hello"
+
+      val y1: Int | String | Boolean = x1
+      val y2: Int | String | Boolean = x2
+      val y3: Int | String | Boolean = x3
+
+      expect(y1).toBe(3)
+      expect(y2).toBe(false)
+      expect(y3).toBe("hello")
+    }
+
+    it("Partial upper bound") {
+      val x: Int | String | Boolean = "hello"
+
+      val x1: AnyVal | String = x
+      val x2: String | AnyVal = x
+
+      expect(x1).toBe("hello")
+      expect(x2).toBe("hello")
+
+      /* Note: the *total* upper bound does not work, because the expected
+       * type is not a | type.
+       */
```

Comment:
Unfortunately, that goes into diverging implicit search (unless you *explicitly* write `toTopBound(x)` instead of just `x`). So it's a bit useless.